### PR TITLE
dev-interpreter

### PR DESCRIPTION
**Pull request by JosDenmark**
Substitution and renaming functions for de Bruijn level indicies

### CODE REVIEW

**Comment by JosDenmark** [comment]
@guardbotmk3 The substitution for `Expr` is a bit weird. A substitution can yield either a variable, a quote, or a process. I don't see a simple way around that. Also, substituting process variables requires that the new process is placed in the appropriate process list. Let me know what you think.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala [hunk 1]
```diff
@@ -12,22 +12,25 @@ case class Par(
     // selects: List[Select],
     evals: List[Eval],
     news: List[New],
-    exprs: List[Expr]
+    exprs: List[Expr],
     // matches: List[Match]
+    id: Option[GPrivate]
```
Comment (KentShikama) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: I was under the assumption that we we're going to use

`case class GPrivate(p: String) extends Ground`

and thus these are just part of the exprs.

Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: We were, but it turns out, it's especially handy to be able to refer directly to an identifier, rather than having it blend with every other expression.

Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: Ok in that case can we remove the

`extends Ground `

then?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        rename(exp, j)
+      },
+      term.id
+    )
+
+  def substitute(term: Quote, env: Env): Quote = Quote(substitute(term.p, env))
+
+  def rename(term: Quote, j: Int): Quote = Quote(rename(term.p, j))
+
+  def substitute(term: Channel, env: Env): Either[Channel, Par] =
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: This should return a Channel, not an either.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    term match {
+      case q @ Quote(_) => Left(substitute(q, env))
+      case ChanVar(b @ _) =>
+        substitute(b, env) match {
+          case Right(Left(quote)) => Left(quote)
+          case Right(Right(par))  => Right(par)
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: This is an error.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+  def substitute(term: Quote, env: Env): Quote = Quote(substitute(term.p, env))
+
+  def rename(term: Quote, j: Int): Quote = Quote(rename(term.p, j))
+
+  def substitute(term: Channel, env: Env): Either[Channel, Par] =
+    term match {
+      case q @ Quote(_) => Left(substitute(q, env))
+      case ChanVar(b @ _) =>
+        substitute(b, env) match {
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: I don't think this should be a substitute call. Just a lookup.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+          chan,
+          term.data map { par =>
+            substitute(par, env)
+          },
+          term.persistent
+        )
+      case Right(par) => throw new Error(s"Cannot send on expression: $par")
+    }
+
+  def rename(term: Send, j: Int): Send =
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: Please group all the renames together, and all the sends together.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+  def rename(term: Channel, j: Int): Channel =
+    term match {
+      case q @ Quote(_)   => rename(q, j)
+      case ChanVar(varue) => ChanVar(rename(varue, j))
+    }
+
+  def substitute(term: Send, env: Env): Send =
+    substitute(term.chan, env) match {
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: substituting into a channel should only ever yield a channel. It shouldn't yield an either.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+  def substitute(term: New, env: Env): New =
+    New(term.count, substitute(term.p, env))
+
+  def rename(term: New, j: Int): New =
+    New(term.count, rename(term.p, j))
+
+  def substitute(term: Receive, env: Env): Receive =
+    Receive(
+      for { (xs, chan) <- term.binds } yield {
+        substitute(chan, env) match {
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: You should be happy that chan substitution can only return a chan. It will make expressions like these simpler.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    )
+
+  def substitute(term: Var, env: Env): Either[Var, Data] =
+    term match {
+      case BoundVar(level) =>
+        env.get(level) match {
+          case Some(Left(quote)) => Right(Left(rename(quote, env.level)))
+          case Some(Right(par))  => Right(Right(rename(par, env.level)))
+          case None =>
+            if (level - env.level < 0) Left(BoundVar(0))
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: This is an error case.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    Receive(
+      for ((xs, channel) <- term.binds) yield {
+        (xs, rename(channel, j))
+      },
+      rename(term.body, j),
+      term.persistent,
+      term.count
+    )
+
+  def substitute(term: Var, env: Env): Either[Var, Data] =
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: Please call this one something different. It makes sense to substitute EVar, but what you're doing here is a little different.
Something like lookupOrDecrement(Var, Env). It makes it more clear that we're at the ground floor.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+            if (level - env.level < 0) Left(BoundVar(0))
+            else Left(BoundVar(level - env.level))
+        }
+      case _ => Left(term)
+    }
+
+  def rename(term: Var, j: Int): Var =
+    term match {
+      case BoundVar(i) =>
+        if (i + j < 0) BoundVar(0)
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: This is an error case. (It's an error to call rename with j < 0)

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+      case ELte(p1: Par, p2: Par)   => ELte(rename(p1, j), rename(p2, j))
+      case EGt(p1: Par, p2: Par)    => EGt(rename(p1, j), rename(p2, j))
+      case EGte(p1: Par, p2: Par)   => EGte(rename(p1, j), rename(p2, j))
+      case EEq(p1: Par, p2: Par)    => EEq(rename(p1, j), rename(p2, j))
+      case ENeq(p1: Par, p2: Par)   => ENeq(rename(p1, j), rename(p2, j))
+      case EAnd(p1: Par, p2: Par)   => EAnd(rename(p1, j), rename(p2, j))
+      case EOr(p1: Par, p2: Par)    => EOr(rename(p1, j), rename(p2, j))
+    }
+
+  def rename(exp: Ground, j: Int): Ground =
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: Does scala really allow this? A Ground is an expression. Shouldn't this conflict with the above?

Even if it does, I'd much rather you split them off in the function above and had a function with a different name for grounds.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+          rename(par, j)
+        })
+      case EMap(xs) =>
+        EMap(for { (par0, par1) <- xs } yield {
+          (rename(par0, j), rename(par1, j))
+        })
+      case g @ _ => g
+    }
+
+  def substitute(exp: EVar, env: Env): Either[Var, Data] =
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: The return type here suggests that you should explicitly handle EVar differently.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 13]
```diff
@@ -1,8 +1,9 @@
 package coop.rchain.rholang.interpreter
 
-import coop.rchain.rholang.syntax.rholang_mercury
 import coop.rchain.rholang.syntax.rholang_mercury.Absyn.{Ground => AbsynGround, _}
 
+//RosetteBaseLanguage
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Comment doesn't seem correct or on target. Please remove.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala [hunk 14]
```diff
@@ -61,16 +64,54 @@ case class Par(
...
+
+  override def toString: String =
+    if (isEmpty) "Nil"
+    else if (isIdentifier) id.get.p.substring(0, 8)
+    else {
+      val collect = List(sends, receives, evals, news, if (id.isDefined) List(id) else Nil)
+      (for { xs <- collect; if xs.nonEmpty } yield {
+        xs.mkString(" | ")
+      }).mkString(" | ")
+    }
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: Please remove the pretty printing portion. It's good work, it just doesn't belong in this PR

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala [hunk 15]
```diff
@@ -61,16 +64,54 @@ case class Par(
...
-  def apply(n: New): Par     = new Par(n)
-  def apply(e: Expr): Par    = new Par(e)
+  def apply(): Par            = new Par()
+  def apply(s: Send): Par     = new Par(s)
+  def apply(r: Receive): Par  = new Par(r)
+  def apply(e: Eval): Par     = new Par(e)
+  def apply(n: New): Par      = new Par(n)
+  def apply(e: Expr): Par     = new Par(e)
+
+  def fromGPrivate: Quote     = Quote(apply(GPrivate(uuid)))
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: from GPrivate belongs in the helper object for Quote.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala [hunk 16]
```diff
@@ -110,15 +160,25 @@ case class Send(chan: Channel, data: List[Par], persistent: Boolean)
 case class Receive(binds: List[(List[Channel], Channel)],
                    body: Par,
                    persistent: Boolean,
-                   count: Int)
+                   count: Int) {
+  override def toString: String =
+    (for { (xs, chan) <- binds; if binds.nonEmpty } yield {
+      xs.mkString(", ") + " <- " + chan.toString
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: Case in point why the pretty printing should have been a separate PR:

if persistent is true, all the <-'s should be <='s

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala [hunk 17]
```diff
@@ -110,15 +160,25 @@ case class Send(chan: Channel, data: List[Par], persistent: Boolean)
...
+}
 
 // Number of variables bound in the new statement.
 // For normalized form, p should not contain solely another new.
 // Also for normalized form, the first use should be level+0, next use level+1
 // up to level+count for the last used variable.
-case class New(count: Int, p: Par)
+case class New(count: Int, p: Par) {
+  override def toString: String =
+    "new " + (0 until count).toList.mkString(", ") + " in { " + p.toString + " }"
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: This suggests that you should either just do new:count as the pretty printing, or that pretty printing needs a level based counter so that it can print new xj, xj_+_1, xj_+_k_-_1.

File: roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala [hunk 18]
```diff
@@ -42,7 +42,9 @@ object VirtualMachine {
   def handleVirtualMachineError(state: VMState): VMState =
     state.ctxt.vmError(state)._2
 
+
```
Comment (guardbotmk3) [inline roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala]: Seems completely unrelated. Please remove.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Env.scala [hunk 19]
```diff
@@ -0,0 +1,26 @@
+package coop.rchain.rholang.interpreter
+
+object Env {
```
[the remark this answers is not in the document]
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Env.scala]: As opposed to a trait with a companion? It's grouped in it's own object so that there's somewhere to put `implicit class DeBruijn`. I did that to extend `HashMap` functionality and keep dot notation.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 4 above)
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: It could be, but we want to delegate variable substitution to another function so that we can substitute variables that aren't wrapped in ChanVar. EVar(Var) is a good example of when that would be useful.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala (hunk 1 above)
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: @KentShikama Yes. Good catch.
@guardbotmk3 Why would a par need multiple par identifiers?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 12 above)
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: This is what my commit message was about. I don't see a way around the fact that a variable may be substituted either for a channel, or for a process, or not at all if there is no binding. I believe the return type is incorrect. It should be `Either[EVar,Data]`, but that makes little difference.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+      },
+      for { eval <- term.evals } yield {
+        substitute(eval, env)
+      },
+      for { neu <- term.news } yield {
+        substitute(neu, env)
+      },
+      term.exprs, /*for { exp <- term.exprs } yield {
+        case e @ EVar(_) =>
+          substitute(e, env) match {
```
[the remark this answers is not in the document]
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: So, if EVars can only be processes, this is totally simple. Is this the case?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 11 above)
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: On the JVM, when the argument is resolved, it's resolved to the most specific type in the hierarchy, so when `rename(GBool(true),i)` is called, `GBool(true)` resolves to `Ground` not `Expr`, thus the second rename is called. 

I can totally split them off, but I'm curious why?

Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: It's fine. I don't like code where I have to know odd pieces of the JVM in order to understand them.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 12 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: What it's substituted for depends solely on where it occurs.
I don't think this should be called "substitute" it breaks symmetry with all the other substitute methods.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Env.scala (hunk 19 above)
[the remark this answers is not in the document]
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Env.scala]: We want to extend `HashMap` with additional functionality. We call a hashmap with these additional functions an environment. Since, you can't directly extend a collection in scala, we introduce an implicit class to group the additional functions. This let's us keep dot syntax. Since an implicit class can't be top level, we put it in the `Env` object.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 5 above)
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: @guardbotmk3 This is what confused me about the substitution rule. If an environment is a mapping from integers to pars, and substituting a channel always yields a channel (no error condition), then the above error doesn't exist. If that's correct, that's fine, I just want to clarify.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Env.scala [hunk 21]
```diff
@@ -0,0 +1,32 @@
+package coop.rchain.rholang.interpreter
+
+import monix.eval.Task
+
+import scala.collection.immutable.HashMap
+
+object Env {
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Env.scala]: I still don't understand this construction.
How is this better than just creating an Env class?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+        }
+      }
+
+    val sends   = for { s <- term.sends } yield substitute(s)
+    val recvs   = for { r <- term.receives } yield substitute(r)
+    val evals   = for { e <- term.evals } yield substitute(e)
+    val news    = for { n <- term.news } yield substitute(n)
+    val matches = for { m <- term.matches } yield substitute(m)
+
+    val freeCount: Int =
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: You shouldn't need to do this. No term with a freecount > 0 should be evaluated, so it also should not be substituted.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 23]
```diff
@@ -0,0 +1,294 @@
...
+
+    def subExp(expxs: List[Expr]): Par =
+      (Par() /: expxs) { (par, expr) =>
+        expr match {
+          case e @ EVar(_) =>
+            subOrDec(e) match {
+              case Left(_e)    => par :+ _e
+              case Right(_par) => par ++ _par
+            }
+          case e @ _ => par :+ substitute(e)
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: It's probably better to go through the list in reverse and prepend to the result list.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 24]
```diff
@@ -0,0 +1,294 @@
...
+      freeCount
+    )
+  }
+
+  def substitute(term: Receive)(implicit env: Env[Par]): Receive = {
+    val _binds = for { (xs, chan) <- term.binds } yield {
+      (xs, substitute(chan))
+    }
+    val _body = substitute(term.body)
+    val freeCount = _body.freeCount + (0 /: _binds) {
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: Same comment here. FreeCount should be 0, and it's just an error if it isn't.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 25]
```diff
@@ -0,0 +1,294 @@
...
+
+  def substitute(term: Match)(implicit env: Env[Par]): Match = {
+    val _value = substitute(term.value)
+    val _cases = for { (_case, par) <- term.cases } yield {
+      (_case, substitute(par))
+    }
+    Match(
+      _value,
+      _cases,
+      _value.freeCount +
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: Same here.

Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: As an alternative, you should see that since you only ever replace bound variables, the free count will never change.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala [hunk 26]
```diff
@@ -45,6 +48,24 @@ case class Par(
     this.copy(exprs = e :: this.exprs, freeCount = this.freeCount + e.freeCount)
   def prepend(m: Match): Par =
     this.copy(matches = m :: this.matches, freeCount = this.freeCount + m.freeCount)
+  def prepend(g: GPrivate): Par =
+    this.copy(id = g :: this.id)
+
+  // Convenience append methods
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: You don't want to append to immutable lists in scala.
It's an anti-pattern.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala [hunk 27]
```diff
@@ -66,14 +87,46 @@ case class Par(
...
 
-  def merge(that: Par) =
-    Par(that.sends ++ sends,
-        that.receives ++ receives,
-        that.evals ++ evals,
-        that.news ++ news,
-        that.exprs ++ exprs,
-        that.matches ++ matches,
-        that.freeCount + freeCount)
+  def ++(that: Par) =
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: This is fine, but again, I'd rather see it in a separate PR

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala [hunk 28]
```diff
@@ -66,14 +87,46 @@ case class Par(
...
+      receives ++ that.receives,
+      evals ++ that.evals,
+      news ++ that.news,
+      exprs ++ that.exprs,
+      matches ++ that.matches,
+      id ++ that.id,
+      that.freeCount + freeCount
+    )
+
+  def procCount: Int =
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: This isn't even used in your PR, please remove.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala [hunk 29]
```diff
@@ -66,14 +87,46 @@ case class Par(
...
+      receives.isEmpty &&
+      evals.isEmpty &&
+      news.isEmpty &&
+      matches.isEmpty &&
+      exprs.isEmpty &&
+      id.isEmpty
+
+  def nonEmpty: Boolean = !isEmpty
+
+  def isIdentifier: Boolean =
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: I don't think this should stay. Any process can be used as an identifier.

Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: Also, it's unused.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala [hunk 30]
```diff
@@ -85,12 +138,16 @@ object Par {
   def apply(e: Expr): Par    = new Par(e)
   def apply(m: Match): Par   = new Par(m)
 
-  implicit def fromSend(s: Send): Par       = apply(s)
-  implicit def fromReceive(r: Receive): Par = apply(r)
-  implicit def fromEval(e: Eval): Par       = apply(e)
-  implicit def fromNew(n: New): Par         = apply(n)
-  implicit def fromExpr(e: Expr): Par       = apply(e)
-  implicit def fromMatch(m: Match): Par     = apply(m)
+  def forge: Par   = apply(GPrivate(uuid))
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: These look like helper methods for your tests. Please put them in the tests.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala [hunk 31]
```diff
@@ -0,0 +1,166 @@
...
+    result should be (Left(BoundVar(0)))
+  }
+}
+
+class ChannelSub extends FlatSpec with Matchers {
+
+  "ChanVar" should "be decremented by environment level" in {
+    val env = HashMap(0 -> Par.forge, 1 -> Par.forge)
+    val result = substitute(chan2)(env)
+    result should be (chan0)
```
Comment (guardbotmk3) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala]: Please define these closer to the tests that use them.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Env.scala (hunk 21 above)
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Env.scala]: Ok. Say I decide to make a class for `Env`. I want that class to have all library functions for `immutable.HashMap` because this implementation is a hash map and they'd be very handy to have. If I try to do `class Env extends HashMap[Int, Par]`, I get an error: `Illegal inheritance from sealed trait HashMap`. I could create a wrapper class that implements the same methods with an internal hash map, but it's a bit sloppy. So, the alternative is to introduce `implicit class DeBruijn`, but I can't do a top level implicit class. I have to put it in a singleton. Think of `Env` as the place where all possible environment implementations live.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 22 above)
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: Doesn't that mean that no terms with free variables should be stored in the environment? If they can, then doing the substitution introduces free variables into the target term.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 24 above)
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: See comment above. If processes with free variables can be subbed in, we need a recount.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala (hunk 30 above)
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: It's not a helper method for tests. It's a method for creating an unforgeable process.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala (hunk 31 above)
Comment (JosDenmark) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala]: They're used globally.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Env.scala (hunk 21 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Env.scala]: You've explained why, but not what the code actually does. I understand that you want an implicit. I don't see how that gets you from A to B. Why can you add a parameter to Env? Etc. This looks like magic to me, and I won't approve it until I understand it. Also, you're going to need to add comments to the code explaining what is going on.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/interpreter.scala [hunk 32]
```diff
@@ -0,0 +1,11 @@
+package coop.rchain.rholang
+
+import scala.collection.immutable.HashMap
+
+package object interpreter {
+
+  type Data = Either[Quote, Par]
+
+  type Env[A] = HashMap[Int, A]
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/interpreter.scala]: If this is related to the Env above, you've made this infinitely harder to understand by putting them in separate files. Find a single home for them, and add comments about what you're doing.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala (hunk 30 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: Please remove it. It doesn't belong in Par. At best it belongs in GPrivate.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala [hunk 33]
```diff
@@ -0,0 +1,164 @@
+package coop.rchain.rholang.interpreter
+
+import org.scalatest.{FlatSpec, Matchers}
+import Substitute._
+import scala.collection.immutable.HashMap
+import SubstituteTest._
+
+object SubstituteTest {
```
Comment (guardbotmk3) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala]: I don't like this at all. All of these are short enough that the tests would be more readable if this object were gone and you used the values directly. Tests are a specification and splitting these out makes them harder to read.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/interpreter.scala (hunk 32 above)
Comment (JosDenmark) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/interpreter.scala]: Definitions in package objects are package visible (you can't import a type alias). What if I rename Env to EnvOps and keep using Env[Par] in the body?

Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/interpreter.scala]: This is closed. Thanks for working with me to make it more clear.
